rustc nightly hangs with cross-crate trait bound issue involving `pub use crate` or `extern crate self`

`rustc 1.93.0-nightly (3d461af2a 2025-11-18)` hangs without producing any output instead of producing an error about an unsatisfied trait bound. The case involves at least two crates, I haven't been able to reduce it further. Here's a short shell script that sets it up by creating the crates in the current directory:

```sh
cargo new --bin user
cargo new --lib library
(cd user && cargo add --path ../library)

cat > user/src/main.rs <<EOF
fn assert_binread<T: library::BinRead>() {}
fn main() {
    assert_binread::<()>();
}
EOF

cat > library/src/lib.rs <<EOF
pub trait BinRead {}
pub struct PosValue<T> {
    pub t: T,
}
impl<T: BinRead> BinRead for PosValue<T> {}
pub use crate as library;
EOF
```

Running those commands and then `cd user && cargo +nightly check`, I observe a hang on:

```
    Building [=============>               ] 1/2: user(bin)
```

Whereas with `cargo check` on 1.91.1 stable it produces the expected error immediately:

```
error[E0277]: the trait bound `(): BinRead` is not satisfied
 --> src/main.rs:3:22
  |
3 |     assert_binread::<()>();
  |                      ^^ the trait `BinRead` is not implemented for `()`
  |
  = help: the trait `BinRead` is implemented for `PosValue<T>`
note: required by a bound in `assert_binread`
 --> src/main.rs:1:22
  |
1 | fn assert_binread<T: library::BinRead>() {}
  |                      ^^^^^^^^^^^^^^^^ required by this bound in `assert_binread`
```

I looked at the open I-hang issues and this does not seem to match any of them. I also tried setting `RUSTFLAGS=-Znext-solver=globally` (as many of them are marked as "fixed-by-next-solver") but the hang still reproduces.

The `library` crate here was reduced from `binrw 0.15.0`. It uses `extern crate self as binrw` similarly to the `pub use crate as library` here — to hazard a guess, maybe that causes some kind of infinite loop related to name resolution.

### Meta

`rustc +nightly --version --verbose`:
```
rustc 1.93.0-nightly (3d461af2a 2025-11-18)
binary: rustc
commit-hash: 3d461af2a23456a2676aadb13b4253c87bdfe28d
commit-date: 2025-11-18
host: x86_64-unknown-linux-gnu
release: 1.93.0-nightly
LLVM version: 21.1.5
```

`RUSTFLAGS=-Ztime-passes cargo +nightly check` output:

```
    Checking library v0.1.0 (/path/to/library)
time:   0.000; rss:   36MB ->   37MB (   +1MB)  parse_crate
time:   0.000; rss:   43MB ->   43MB (   +0MB)  crate_injection
time:   0.003; rss:   43MB ->   53MB (  +10MB)  expand_crate
time:   0.004; rss:   43MB ->   53MB (  +10MB)  macro_expand_crate
time:   0.000; rss:   53MB ->   53MB (   +0MB)  AST_validation
time:   0.000; rss:   53MB ->   54MB (   +0MB)  finalize_macro_resolutions
time:   0.000; rss:   54MB ->   54MB (   +1MB)  late_resolve_crate
time:   0.000; rss:   54MB ->   54MB (   +0MB)  resolve_check_unused
time:   0.001; rss:   53MB ->   54MB (   +1MB)  resolve_crate
time:   0.000; rss:   54MB ->   55MB (   +0MB)  write_dep_info
time:   0.000; rss:   55MB ->   55MB (   +0MB)  complete_gated_feature_checking
time:   0.000; rss:   56MB ->   57MB (   +0MB)  looking_for_entry_point
time:   0.000; rss:   56MB ->   57MB (   +0MB)  misc_checking_1
time:   0.000; rss:   57MB ->   57MB (   +0MB)  coherence_checking
time:   0.000; rss:   57MB ->   57MB (   +0MB)  type_check_crate
time:   0.000; rss:   58MB ->   58MB (   +0MB)  lint_checking
time:   0.002; rss:   57MB ->   58MB (   +1MB)  misc_checking_3
time:   0.000; rss:   58MB ->   58MB (   +0MB)  generate_crate_metadata
time:   0.000; rss:   58MB ->   59MB (   +1MB)  codegen_crate
time:   0.000; rss:   59MB ->   60MB (   +0MB)  encode_query_results
time:   0.000; rss:   59MB ->   60MB (   +0MB)  incr_comp_serialize_result_cache
time:   0.003; rss:   59MB ->   60MB (   +1MB)  incr_comp_persist_result_cache
time:   0.003; rss:   59MB ->   60MB (   +1MB)  serialize_dep_graph
time:   0.028; rss:   28MB ->   49MB (  +21MB)  total
    Checking user v0.1.0 (/path/to/user)
time:   0.000; rss:   36MB ->   37MB (   +1MB)  parse_crate
time:   0.000; rss:   43MB ->   43MB (   +0MB)  crate_injection
time:   0.003; rss:   43MB ->   53MB (  +10MB)  expand_crate
time:   0.003; rss:   43MB ->   53MB (  +10MB)  macro_expand_crate
time:   0.000; rss:   53MB ->   53MB (   +0MB)  AST_validation
time:   0.000; rss:   53MB ->   53MB (   +0MB)  finalize_macro_resolutions
time:   0.000; rss:   53MB ->   54MB (   +0MB)  late_resolve_crate
time:   0.000; rss:   54MB ->   54MB (   +0MB)  resolve_check_unused
time:   0.001; rss:   53MB ->   54MB (   +1MB)  resolve_crate
time:   0.000; rss:   54MB ->   54MB (   +0MB)  write_dep_info
time:   0.000; rss:   54MB ->   54MB (   +0MB)  complete_gated_feature_checking
time:   0.000; rss:   56MB ->   56MB (   +0MB)  looking_for_entry_point
time:   0.001; rss:   56MB ->   57MB (   +1MB)  misc_checking_1
time:   0.001; rss:   57MB ->   62MB (   +4MB)  coherence_checking
    Building [=============>               ] 1/2: user(bin)
```

--- a/compiler/rustc_trait_selection/src/error_reporting/traits/fulfillment_errors.rs
+++ b/compiler/rustc_trait_selection/src/error_reporting/traits/fulfillment_errors.rs
@@ -1,12 +1,13 @@
 // ignore-tidy-filelength
 use core::ops::ControlFlow;
 use std::borrow::Cow;
+use std::collections::hash_set;
 use std::path::PathBuf;
 
 use rustc_abi::ExternAbi;
 use rustc_ast::ast::LitKind;
 use rustc_ast::{LitIntType, TraitObjectSyntax};
-use rustc_data_structures::fx::FxHashMap;
+use rustc_data_structures::fx::{FxHashMap, FxHashSet};
 use rustc_data_structures::unord::UnordSet;
 use rustc_errors::codes::*;
 use rustc_errors::{
@@ -1956,11 +1957,17 @@ impl<'a, 'tcx> TypeErrCtxt<'a, 'tcx> {
                         if self.tcx.visibility(did).is_accessible_from(body_def_id, self.tcx) {
                             // don't suggest foreign `#[doc(hidden)]` types
                             if !did.is_local() {
-                                while let Some(parent) = parent_map.get(&did) {
+                                let mut previously_seen_dids: FxHashSet<DefId> = Default::default();
+                                previously_seen_dids.insert(did);
+                                while let Some(&parent) = parent_map.get(&did)
+                                    && let hash_set::Entry::Vacant(v) =
+                                        previously_seen_dids.entry(parent)
+                                {
                                     if self.tcx.is_doc_hidden(did) {
                                         return false;
                                     }
-                                    did = *parent;
+                                    v.insert();
+                                    did = parent;
                                 }
                             }
                             true

--- a/compiler/rustc_trait_selection/src/lib.rs
+++ b/compiler/rustc_trait_selection/src/lib.rs
@@ -17,6 +17,7 @@
 #![feature(associated_type_defaults)]
 #![feature(box_patterns)]
 #![feature(default_field_values)]
+#![feature(hash_set_entry)]
 #![feature(if_let_guard)]
 #![feature(iter_intersperse)]
 #![feature(iterator_try_reduce)]

--- a/tests/ui/suggestions/auxiliary/hidden-struct.rs
+++ b/tests/ui/suggestions/auxiliary/hidden-struct.rs
@@ -33,3 +33,6 @@ impl Marker for hidden::Foo {}
 impl Marker for hidden1::Bar {}
 impl Marker for Baz {}
 impl Marker for Quux {}
+
+
+pub use crate as library;

--- a/tests/ui/suggestions/dont-suggest-foreign-doc-hidden.rs
+++ b/tests/ui/suggestions/dont-suggest-foreign-doc-hidden.rs
@@ -21,7 +21,12 @@ pub fn test4(_: Quux) {}
 
 fn test5<T: hidden_struct::Marker>() {}
 
+fn test6<T: hidden_struct::library::Marker>() {}
+
 fn main() {
     test5::<i32>();
+    //~^ ERROR [E0277]
+
+    test6::<i32>();
     //~^ ERROR [E0277]
 }

--- a/tests/ui/suggestions/dont-suggest-foreign-doc-hidden.stderr
+++ b/tests/ui/suggestions/dont-suggest-foreign-doc-hidden.stderr
@@ -38,7 +38,7 @@ LL + use hidden_struct::Quux;
    |
 
 error[E0277]: the trait bound `i32: Marker` is not satisfied
-  --> $DIR/dont-suggest-foreign-doc-hidden.rs:25:13
+  --> $DIR/dont-suggest-foreign-doc-hidden.rs:27:13
    |
 LL |     test5::<i32>();
    |             ^^^ the trait `Marker` is not implemented for `i32`
@@ -59,7 +59,29 @@ note: required by a bound in `test5`
 LL | fn test5<T: hidden_struct::Marker>() {}
    |             ^^^^^^^^^^^^^^^^^^^^^ required by this bound in `test5`
 
-error: aborting due to 5 previous errors
+error[E0277]: the trait bound `i32: Marker` is not satisfied
+  --> $DIR/dont-suggest-foreign-doc-hidden.rs:30:13
+   |
+LL |     test6::<i32>();
+   |             ^^^ the trait `Marker` is not implemented for `i32`
+   |
+help: the following other types implement trait `Marker`
+  --> $DIR/auxiliary/hidden-struct.rs:31:1
+   |
+LL | impl Marker for Option<u32> {}
+   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^ `Option<u32>`
+...
+LL | impl Marker for Baz {}
+   | ^^^^^^^^^^^^^^^^^^^ `Baz`
+LL | impl Marker for Quux {}
+   | ^^^^^^^^^^^^^^^^^^^^ `Quux`
+note: required by a bound in `test6`
+  --> $DIR/dont-suggest-foreign-doc-hidden.rs:24:13
+   |
+LL | fn test6<T: hidden_struct::library::Marker>() {}
+   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `test6`
+
+error: aborting due to 6 previous errors
 
 Some errors have detailed explanations: E0277, E0412.
 For more information about an error, try `rustc --explain E0277`.

